### ISSUE TITLE
[Versal BELAttr] Parsing issue

### ISSUE THREAD

**Issue by nqdtan** [Xilinx/RapidWright#912]
Hi,

I got the following output when reading the attached design checkpoint

>==============================================================================
==                    Reading DCP: socket_standalone.dcp                    ==
==============================================================================
 XML Parse & Device Load:     0.902s
    (Read PhysDB Caches):            (    0.272s)
   (Read PhysDB Routing):            (    2.100s)
            (EDIF Parse):            (    2.420s)
 (Read PhysDB Placement):            (    3.333s)
(Read PhysDB Finalizing):            (    0.336s)
   Read Netlist & PhysDB:     3.950s
ERROR: Couldn't find net design_1_i/socket/socket_top_0/inst/rg_cl_wrapper/custom_logic/custom_logic_inst/grp_cl_cholesky_Pipeline_VITIS_LOOP_287_3_fu_866/sdiv_64ns_64ns_64_69_1_U54/cl_cholesky_sdiv_64ns_64ns_64_69_1_ip_u/U0/i_synth/i_nonzero_fract.i_synth/i_algo_r2_or_lutmult.i_nonzero_fract.i_sdivider/i_sdivider.divider_blk/div_loop[10].num_stages.numerator_gen.del_numer/aclk in parsing BELAttrDB

RW was able to finish reading the checkpoint. However, it seems that not all BEL attributes were parsed successfully. Might I ask if you could take a look to see what was the reason?

Thanks!

[belattr_issue](https://drive.google.com/file/d/1e-wRDAh5GrxlTzqC7NCmWN1JPPOGZqqg/view?usp=sharing)

**Comment by clavin-xlnx** [comment]
This design (`socket.dcp`) has some encrypted cells in it.  When writing out the EDIF file from Vivado, it should be accompanied by one or possibly many `*.edn` files which contain the encrypted contents of the protected cells.  Thus, the EDIF file will have some black boxes in it when loaded in RapidWright.  RapidWright can generally handle these black boxes without a problem, especially if the design is placed and routed already.  

This issue is due to some poor assumptions in the `rapidwright-api-lib` and was trying to find the net (which falls into an encrypted cell definition) in the EDIF netlist.  I have created a fix/workaround that should allow this to be a non-issue going forward and will be in the 2023.2.1 release (I can have the fix in the next release candidate probably in the next day or two).  

If you are going to write out the DCP, be sure to watch for the message stating that in order to read it back into Vivado, you'll need to `source` the Tcl script that gets generated rather than use the typical `open_checkpoint` command to fully load it back into Vivado.

**Comment by nqdtan** [comment]
Actually, I have a follow-up question. When I iterate through all the BEL attributes to check their corresponding nets as follows (using the same input design checkpoint):

```java
    for (SiteConfig siteConfig : design.getBELAttrs().values()) {                
      for (Map.Entry<BEL, Map<String, BELAttr>> entry1 : siteConfig.getBELAttributes().entrySet()) {
        BEL bel = entry1.getKey();                                               
        for (Map.Entry<String, BELAttr> entry2 : entry1.getValue().entrySet()) { 
          String name = entry2.getKey();                                         
          BELAttr belAttr = entry2.getValue();                                   
                                                                                 
          System.out.println(name + " " + belAttr + " " + belAttr.getNet());                                                                                                           
        }                                                                        
      }                                                                          
    }
```

I got several `null` results:

>IMUX_CLK_MODE design_1_i/mbufgce_primitive_0/inst/clk_out_o1                     
CLK_DLY_VAL design_1_i/mbufgce_primitive_0/inst/clk_out_o1                       
FF_CLK_DUAL design_1_i/mbufgce_primitive_0/inst/clk_out_o1                       
FF_CLK_EN design_1_i/mbufgce_primitive_0/inst/clk_out_o1                         
IMUX_CLK_MODE design_1_i/mbufgce_primitive_0/inst/clk_out_o1                     
CLK_DLY_VAL design_1_i/mbufgce_primitive_0/inst/clk_out_o1                       
FF_CLK_DUAL design_1_i/mbufgce_primitive_0/inst/clk_out_o1                       
FF_CLK_EN design_1_i/mbufgce_primitive_0/inst/clk_out_o1                         
IMUX_CLK_MODE null

It is not a critical issue for me at the moment, since I can safely assume that all these BEL attributes just refer to one (or two) specific clock net in my design, and I can move on with that. However, it is not ideal that the tool returns several `null`s in this case. Perhaps this issue is also caused by the encrypted cells in the design?

**Comment by clavin-xlnx** [comment]
> > IMUX_CLK_MODE null
> 
> It is not a critical issue for me at the moment, since I can safely assume that all these BEL attributes just refer to one (or two) specific clock net in my design, and I can move on with that. However, it is not ideal that the tool returns several `null`s in this case. Perhaps this issue is also caused by the encrypted cells in the design?

Thanks for bringing this up!  There was an issue with the fix (incorrect assignment), I have it corrected now and will push a new change in the next release candidate.

**Comment by clavin-xlnx** [comment]
This should be resolved in the 2023.2.1 release.
